### ISSUE TITLE
[Serializer] Type error in appDevDebugProjectContainer when using serializer

### ISSUE THREAD

**Issue by jverdeyen** [symfony/symfony#17676]
Symfony 2.8.1
PHP 7.0.2.1 (debian)

I've enabled the Symonfy serializer.

I'm using the `"friendsofsymfony/rest-bundle": "^1.7"` with the following config:

```yaml
fos_rest:
    param_fetcher_listener: force
    routing_loader:
        include_format:       false
    view:
        mime_types:
            json: ['application/json', 'application/json;version=1.0', 'application/json;version=1.1', 'application/json;version=1.2']
        view_response_listener: 'force'
        formats:
            xml:  true
            json: true
        templating_formats:
            html: true
    format_listener:
        rules:
            - { path: ^/, priorities: [ html, json, xml ], fallback_format: ~, prefer_extension: true }
        media_type:
            version_regex: '/(v|version)=(?P<version>[0-9\.]+)/'
    exception:
        codes:
            'Symfony\Component\Routing\Exception\ResourceNotFoundException': 404
            'Doctrine\ORM\OptimisticLockException': HTTP_CONFLICT
        messages:
            'Symfony\Component\Routing\Exception\ResourceNotFoundException': true
    allowed_methods_listener: true
    access_denied_listener:
        json: true
    body_listener: true
```

And this is a part of my controller action:

```php
public function getNewsAction(Request $request)
    {
       ...
        $news = [generator news]..
       ...

        $view = $this->view($news);
        return $this->handleView($view, 200);
    }
```

I'm trying to debug the following error:

```
An Exception was thrown while handling: Type error: Argument 1 passed to 
FOSRestBundleSerializerSymfonySerializerAdapter_0000000052c9fc6b000000005a32441d7f491784
a700368952f1ec7813de5f79::__construct() must implement interface 
Symfony\Component\Serializer\SerializerInterface, instance of Closure given, called in 
/tmpfs/dev/cache/appDevDebugProjectContainer.php on line 2766
```

Even when using the JMSSerializer I get:

```
An Exception was thrown while handling: Type error: Argument 1 passed to 
FOSRestBundleSerializerJMSSerializerAdapter_000000003e79141b0000000013a926a47f491784a70
0368952f1ec7813de5f79::__construct() must implement interface JMS\Serializer\SerializerInterface, 
instance of Closure given, called in /tmpfs/dev/cache/appDevDebugProjectContainer.php on line 2789
```

This is the generated code:

```php
/**
     * Gets the 'fos_rest.serializer' service.
     *
     * This service is shared.
     * This method always returns the same instance of the service.
     *
     * @param bool    $lazyLoad whether to try lazy-loading the service with a proxy
     *
     * @return \FOS\RestBundle\Serializer\JMSSerializerAdapter A FOS\RestBundle\Serializer\JMSSerializerAdapter instance.
     */
    public function getFosRest_SerializerService($lazyLoad = true)
    {
        if ($lazyLoad) {
            $container = $this;

            return $this->services['fos_rest.serializer'] = new FOSRestBundleSerializerJMSSerializerAdapter_000000003e79141b0000000013a926a47f491784a
700368952f1ec7813de5f79(
                function (&$wrappedInstance, \ProxyManager\Proxy\LazyLoadingInterface $proxy) use ($container) {
                    $wrappedInstance = $container->getFosRest_SerializerService(false);

                    $proxy->setProxyInitializer(null);

                    return true;
                }
            );
        }

        return new \FOS\RestBundle\Serializer\JMSSerializerAdapter($this->get('jms_serializer'));
    }
```

Any idea where I should start looking for this problem?
I've also seen it in another project (unsolved).

**Comment by xabbuh** [comment]
Can you show the output of `composer show -i friendsofsymfony/rest-bundle` please?

**Comment by jverdeyen** [comment]
```
root@525db9f02719:/var/www/app# composer show -i friendsofsymfony/rest-bundle
name     : friendsofsymfony/rest-bundle
descrip. : This Bundle provides various tools to rapidly develop RESTful API's with Symfony
keywords : rest
versions : * 1.8.x-dev
type     : symfony-bundle
license  : MIT License (MIT) (OSI approved) https://spdx.org/licenses/MIT.html#licenseText
source   : [git] https://github.com/FriendsOfSymfony/FOSRestBundle.git 95dbe97d682e7c3f50ee53947c348c76350a0153
dist     : [zip] https://api.github.com/repos/FriendsOfSymfony/FOSRestBundle/zipball/95dbe97d682e7c3f50ee53947c348c76350a0153 95dbe97d682e7c3f50ee53947c348c76350a0153
names    : friendsofsymfony/rest-bundle

autoload
psr-0
FOS\RestBundle => .

requires
doctrine/inflector ~1.0
php ^5.3.9|~7.0
psr/log ~1.0
symfony/framework-bundle ~2.3|~3.0
symfony/http-kernel ^2.3.24|~3.0
willdurand/jsonp-callback-validator ~1.0
willdurand/negotiation ~1.2

requires (dev)
jms/serializer ~0.13|~1.0
jms/serializer-bundle ~0.12|~1.0
phpoption/phpoption ~1.1.0
sensio/framework-extra-bundle ~2.0|~3.0
sllh/php-cs-fixer-styleci-bridge ^1.3
symfony/browser-kit ~2.3|~3.0
symfony/dependency-injection ~2.3|~3.0
symfony/form ~2.3|~3.0
symfony/phpunit-bridge ~2.7|~3.0
symfony/security ~2.3|~3.0
symfony/serializer ~2.3|~3.0
symfony/validator ~2.3|~3.0
symfony/yaml ~2.3|~3.0

suggests
jms/serializer-bundle Add support for advanced serialization capabilities, recommended, requires ~0.12||~1.0
sensio/framework-extra-bundle Add support for route annotations and the view response listener, requires ~3.0
symfony/serializer Add support for basic serialization capabilities and xml decoding, requires ~2.3
symfony/validator Add support for validation capabilities in the ParamFetcher, requires ~2.3

conflicts
jms/serializer <0.12
jms/serializer-bundle <0.11
symfony/validator >=2.5.0,<2.5.5
```

**Comment by jmcclell** [comment]
I'm having the same issue.

**Comment by jverdeyen** [comment]
@jmcclell: Tested on 1.7.7 of the FOSRestBundle and this is working.

**Comment by jakzal** [comment]
@jverdeyen isn't this an issue with the 1.8-dev version of FOSRestBundle?

**Comment by xabbuh** [comment]
The feature making use of the lazy services was introduced in the 1.8 dev branch, but it doesn't look that we do anything strange with the lazy services: https://github.com/FriendsOfSymfony/FOSRestBundle/commit/b264edc62346881ef7cf44008845bb2a42281693#diff-6fb4ad4aa9f78da8ab275d84ae7ea143

I try to have a look at this this weekend and figure out if this is an issue in Symfony or in FOSRestBundle.

**Comment by jmcclell** [comment]
https://github.com/jmcclell/symfony-testcase-17676

Here's a basic test case. It actually isn't FOSRestBundle at all. In my case, because I've been using minimum-stability = dev while doing some upgrading and testing of new components, I had updated to the latest Doctrine Migrations which uses the 2.0.x branch of the Proxy package. 

The bug appears to be with the Proxy package itself, or, perhaps, Symfony's usage of it. I'll be backing to stable versions and avoiding the 2.0.x branch, so that will fix things for me for now, but it looks like this will need addressed from one end or the other at some point.

**Comment by boast** [comment]
The proxy bridge seems to handle the lazy-proxy generation wrong (as now a Closure will be generated) thus he line here: https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php#L80

should be changed from:
`$instantiation new $proxyClass`
to
`$instantiation $proxyClass::staticProxyConstructor`
and everything should be working fine again. However, this requires the 2.0 branch of the ocramius/proxy-manager package.

**Comment by stof** [comment]
@Ocramius could you have a look at this issue please ?

**Comment by stof** [comment]
@boast on a side note, we don't officially support 2.0 versions of ProxyManager, so getting errors when using may totally be expected.

**Comment by boast** [comment]
I know that :) Maybe the BC-break was unintentional? As mentioned before,
doctrine/migrations allowed the version 2 and in symfonys composer.json,
the proxy-manager is only a dev-requirement.

**Comment by Ocramius** [comment]
Yes, the BC break was intentional here, although I am not sure if I missed it in the upgrade docs. Needs either a context switch (check for PM version) or relying on ProxyManager's factories (which encapsulate this behaviour)

**Comment by Ocramius** [comment]
Ah, no, I documented it, but it's only documented as feature: https://github.com/Ocramius/ProxyManager/blob/6e15877fc03cc211ce798d497a21208a44f7b726/CHANGELOG.md#constructors-in-proxies-are-not-replaced-anymore :-\

**Comment by Ocramius** [comment]
Yes, I think the quickest fix here would be to add a conditional around https://github.com/symfony/symfony/blob/1b857996029b1aecfa393b7b6a92653cc3873363/src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php#L80 to either use `new $proxyClassName` or `$proxyClassName::staticProxyConstructor()`. Since no reflection of the proxy is available there, the quickest solution would be to check for the ProxyManager existence of the [`StaticProxyConstructor`](https://github.com/Ocramius/ProxyManager/blob/6e15877fc03cc211ce798d497a21208a44f7b726/src/ProxyManager/ProxyGenerator/LazyLoading/MethodGenerator/StaticProxyConstructor.php) method generator.

**Comment by Ocramius** [comment]
I provided a fix via #17919
